README.md
# approval-path-visualizer

Read an approval path — steps, approvers, branch conditions, timeouts and exception exits — check
the paths through it, and draw a standalone **SVG** or **HTML** diagram of what it found.

- **Repository:** [edilec/approval-path-visualizer](https://github.com/edilec/approval-path-visualizer)
- **Area:** Automation & Workflows
- **License:** MIT
- Node ESM, `node >= 22`, **no dependencies** — runtime or development. Node built-ins only.

Two questions are hard to answer by reading an approval definition and easy to answer by walking it:

| Rule | What it means |
| --- | --- |
| `node-unreachable` | No path from the start step reaches this one. If it is an approval step, its approvers are never asked — the review everyone believes is happening is not. |
| `approval-cycle` | Two or more steps can send a request round to each other forever. Nobody is refusing it; it just never arrives. |

And one property matters more than either, because the tool writes markup: **every label, approver
name, condition and reason in the input is untrusted text**, and all of them end up inside a
generated document. A step labelled `<script>alert(1)</script>` renders as the words
`<script>alert(1)</script>`, drawn in a box, doing nothing.

## Install

```sh
npm install approval-path-visualizer
```

Or run it from a checkout with `node bin/approval-path-visualizer.mjs`.

## Use

```sh
approval-path-visualizer \
  --root examples/clean \
  --graph approval.json \
  --out build/approval-path.svg
```

```
capital purchase approval (approval.json): 5 approval step(s), 3 outcome(s), 6 approver(s), status pass.
paths: 11 edge(s), 5 timeout(s), 2 exception exit(s), 0 unreachable step(s), 0 cycle(s).
diagram: build/approval-path.svg
```

The deliberately broken example in `examples/broken` exits `1` and says why:

```
ERROR   approval.json/nodes/security-review node-unreachable No path from the start step "intake"
        reaches this node. Its approvers (security-officer, data-protection-officer) are never asked.
ERROR   approval.json/nodes/compliance-review approval-cycle Approval cycle: compliance-review ->
        legal-review -> compliance-review. A request that enters it can be sent round forever.
ERROR   approval.json/nodes/vendor-hold path-dead-end Approval step has no edge, timeout or exception
        exit, so a request that arrives here can never leave.
```

`--json` emits the machine-readable report on stdout, `--format html` writes a page with the diagram
above the findings, and `--help` prints the full flag list.

### As a library

```js
import { visualizeApprovalPath, writeDiagram } from 'approval-path-visualizer'

const { report, diagram, graph, analysis } = await visualizeApprovalPath({
  root: 'examples/clean',
  graph: 'approval.json',
  format: 'svg',
})
```

`visualizeApprovalPath` reads; it never writes. `diagram` is a string, and where it goes is the
caller's decision — `writeDiagram` is offered for convenience and refuses nothing on its own, which
is why the CLI resolves the destination with `resolveDiagramDestination(root, out, graphPath)`
first. Pass that third argument: it is what refuses a destination that is the graph file reached by
another name.

The pure pieces are exported too: `compileGraph`, `analyseGraph`, `renderDiagram`, `layoutGraph`,
`findCycles`, `layerFromStart`, `canonicalCycle`, `escapeMarkup`, `markupText`, `parseDuration`.

## The graph

One JSON file, every key closed. `docs/approval-path-rules.md` holds the full schema; the shape is:

```json
{
  "schemaVersion": "1",
  "name": "capital purchase approval",
  "start": "intake-review",
  "nodes": [
    {
      "id": "intake-review",
      "kind": "approval",
      "label": "Intake review",
      "approvers": ["requesting-team-lead"],
      "timeout": { "after": "P2D", "to": "escalate-to-operations" },
      "edges": [
        { "to": "finance-review", "condition": "amount >= 10000" },
        { "to": "approved", "condition": "amount < 10000" }
      ],
      "exceptions": [{ "to": "withdrawn", "reason": "requester withdrew the request" }]
    },
    { "id": "approved", "kind": "outcome", "label": "Approved", "outcome": "approved" }
  ]
}
```

An unknown key is a finding rather than something quietly ignored: a misspelled `edges` removes a
path, and a removed path is exactly the kind of real failure a typo must not be able to turn green.

## Escaping is the security property

Two passes run on every untrusted value, in this order, and there is one function that does both:

1. **Sanitise** — remove C0 (U+0000–U+001F), DEL, C1 (U+0080–U+009F), U+2028, U+2029 and the bidi
   and isolate controls (U+200E, U+200F, U+202A–U+202E, U+2066–U+2069). U+0085 and U+009B forge
   lines in a report; U+202E reverses what is displayed after it. Escaping does nothing about any of
   them, and this applies to **every** value that reaches output — ids, pointers, paths, messages
   and evidence, not only an excerpt field.
2. **Escape** — `&`, `<`, `>`, `"` and `'`, ampersand first. Both quote styles, so the same function
   is safe inside an attribute; `>` as well as `<`, which is what stops a label ending `]]>` from
   closing a CDATA section in a consuming document and `-->` from ending a comment.

A control character in an **identifier** is refused outright instead of cleaned: a node id that
prints differently from the id that was compared is one nobody can audit.

Sanitising is not what keeps the graph's *content* out of a report. `JSON.parse` names the input
in one of its two failure messages — `Unexpected token 'A', "AKIAIOSFODNN7EXAMPLE" is not valid
JSON`, the whole document when it is short — and that quote sits at the front of the message, where
neither stripping controls nor cutting from the end removes it. The `graph-not-json` evidence is
therefore built from the failure's **position, line and column only**; the quoted input never leaves
the function that reads it.

## Exit codes

| Code | Meaning | stdout |
| ---: | --- | --- |
| `0` | every step is reachable and every path ends at an outcome | the report |
| `1` | the graph was read and failed the check | the report |
| `2` | invalid usage or configuration | **empty** |
| `2` | evidence missing, undecodable or bounded out | an `incomplete` report |

A graph file that was named but is not there is missing evidence, not a configuration error: it is
reported as `graph-unreadable` in an `incomplete` report. stdout is empty only when the run never had
a subject at all — an unusable flag, an unreadable root, a path that resolves outside it.

stdout carries the report and nothing else; diagnostics go to stderr. A consumer that pipes stdout
must handle it being empty on exit 2 — emitting a fake report for a run that never started would be
worse.

## Guarantees

Each of these has a test that fails when the line enforcing it is removed. Each was also checked by
deleting that line, watching the suite go red, and putting it back.

- **A label naming an element renders as text.** Twelve payloads through seven arrival sites — a
  node identifier and an outcome value among them — in both formats, asserted against the document's
  element and attribute vocabulary rather than against what an escape function returns.
- **The input is never written to.** The graph file is byte-identical, to the millisecond, after any
  run; the diagram destination is refused if it resolves inside the input root, refused through a
  symlinked parent, and refused when it is the graph file itself under a second name. `realpath`
  resolves a symbolic link, but a hard link has no target — two names for one inode resolve to two
  different paths, so identity is decided on `(device, inode)` and never on the resolved path.
- **A verdict is withheld rather than guessed.** `node-unreachable`, `approval-cycle` and
  `outcome-unreachable` are reported only when the traversal that decides them ran to completion. A
  bounded-out or untraceable walk is `incomplete`, which is not a pass and not a failure.
- **`pass` with `checked: 0` is not reachable.** A graph that yielded no step is `incomplete`.
- **Every documented limit is enforced**, reported by name, and makes the run `incomplete` — never a
  quietly shorter diagram. Nothing is truncated in silence: a value above `maxLabelLength` is cut
  with an ellipsis *and* a finding, and the renderer wraps rather than cuts. Nothing past a limit is
  read, either: the counts in the summary stop where the finding says the reading stopped.
- **A display bound decides nothing.** `maxLabelLength` bounds what the diagram draws; the outcome
  vocabulary check and the comparisons behind `condition-duplicate`, `edge-duplicate` and
  `approvers-duplicate` all run on the whole sanitised value, so lowering the bound cannot invent a
  finding about a graph that is fine.
- **Every finding's severity comes from one frozen table**, asserted against the documented catalog
  in both directions — and, because agreeing declarations can be edited together, all forty-one
  rules are pinned again by behaviour in a file that imports no table, reads no catalog and holds no
  map of expected values. Each expectation there is a literal at its assertion: status, exit code

CHANGELOG.md
# Changelog

All notable changes to this project are documented in this file.

## Unreleased

### Fixed

- `--out` accepted a destination that destroyed a file the tool was never asked
  to touch, and reported success while doing it. `resolveDiagramDestination`
  resolved the destination through its nearest existing ancestor, so a
  **symbolic link at `--out`** was resolved by that very call and the diagram
  was written through it: measured here, a 14-byte file outside the root became
  an 11117-byte SVG while the run exited 0 and printed `diagram written`. A
  link whose target did not exist yet created the file out there instead, and a
  **symlinked parent directory** did the same thing one level up with nothing
  comparing the resolved parent against a permitted root, because there was no
  permitted root. Both are now refused by `assertWritableDestination` before
  anything is opened; the hard-link refusal that was already here is now made
  by the same guard, by device and inode. A refused destination is a
  configuration error: exit 2, empty stdout. `test/destination.test.mjs` has
  one case per hole and one per allowed shape, and every check in the guard has
  a mutation that reddens the test named for it.

- `graph-not-json` no longer reproduces the graph file inside its own evidence.
  `JSON.parse` reports a failure either by position or by quoting the input
  back — `Unexpected token 'A', "AKIAIOSFODNN7EXAMPLE" is not valid JSON`, which
  is the whole document when the document is short — and that quote sits at the
  front of the message, where sanitising and excerpting both leave it intact. A
  graph file short enough to be only a credential was therefore published in
  full by the finding that failed to read it. Evidence is now built from the
  position, line and column alone.

### Added

- `--out-root`, declaring the tree `--out` may resolve inside. It defaults to
  the working directory and has no meaning without `--out`;
- `assertWritableDestination` and `DestinationError`, exported for a caller
  writing its own destination logic;
- an explicit approval graph — steps, approvers, branch conditions, timeouts and
  exception exits — compiled from data, with every unknown key at every level
  refused rather than ignored, and every structurally defective node dropped
  whole rather than half-read;
- the two verdicts the tool exists for: an approval step no path from the start
  reaches (`node-unreachable`, whose message names the approvers who are never
  asked) and an approval cycle (`approval-cycle`, canonically rotated so one
  loop is reported once however it is entered), both decided over edges,
  timeouts and exception exits alike;
- the verdicts that follow from the same walk: an approval step from which no
  outcome is reachable, a step a request can never leave, a graph with no
  outcome node at all, and an edge, timeout or exception exit that names a step
  nothing declares;
- **withheld verdicts.** `node-unreachable`, `approval-cycle` and
  `outcome-unreachable` are reported only when the traversal that decides them
  ran to completion. An unknown start step, the depth bound or the work budget
  stopping the walk withholds all three and makes the run `incomplete`: a step
  the walk never visited is not evidence that nothing reaches it, and reporting
  it as one would be a fabricated finding in a report whose only value is that
  it can be trusted;
- a standalone SVG or HTML diagram of the compiled graph, laid out by integer
  arithmetic — a shortest-path layer index, a character count, a fixed box
  width — so no font is measured and the same graph renders to the same bytes
  everywhere. Steps no path reaches are drawn in a band of their own, steps on
  a cycle are marked, and each box is one `aria-label`-ed group;
- escaping as the security property, not a presentation detail: `&`, `<`, `>`,
  `"` and `'` on every interpolation, so a label naming a script element, an
  attribute break, a comment opener or `]]>` renders as the text of one;
- sanitisation of every untrusted string that reaches output — identifiers,
  pointers, paths and messages as well as `evidence`, and a CLI diagnostic on
  its way to stderr — covering C0 and DEL, the whole C1 range (U+0085 NEL
  forges a line of its own, U+009B is the 8-bit CSI), U+2028 and U+2029, and
  the bidi and isolate controls U+200E, U+200F, U+202A–U+202E and U+2066–U+2069,
  which are also refused inside an identifier. Ordinary right-to-left letters
  are untouched: they carry their own direction and need no override;
- strict UTF-8 decoding with `TextDecoder('utf-8', { fatal: true })`, so

src/index.mjs
/**
 * approval-path-visualizer
 *
 * Reads one approval graph -- steps, approvers, branch conditions, timeouts
 * and exception exits -- checks the paths through it, and renders a standalone
 * SVG or HTML diagram of what it found.
 *
 * Three properties are structural rather than incidental:
 *
 * 1. **The input is opened read-only.** The diagram is a derived artifact
 *    written to a destination that is refused if it resolves inside the input
 *    root -- or if it is the graph file itself under a second name, which a
 *    hard link is and no path comparison can see -- so a run can never
 *    overwrite the graph it was asked to draw, and there is no auto-fix of any
 *    kind.
 * 2. **Every untrusted string is sanitised and escaped before it enters the
 *    output.** Labels, approver names, conditions, reasons and ids all come
 *    from a file this tool did not write and all end up inside markup. A label
 *    spelling `<script>` renders as the text `<script>`.
 * 3. **A verdict is withheld rather than guessed.** "No path reaches this
 *    step" is a serious accusation; it is only made when the traversal that
 *    decides it ran to completion. A bounded-out walk is `incomplete`, which
 *    is not a pass and not a failure but a statement that the tool did not
 *    find out.
 */

import { mkdir, readFile, realpath, stat, writeFile } from 'node:fs/promises'
import { basename, dirname, isAbsolute, join, resolve, sep } from 'node:path'

import { analyseGraph, compileGraph, outgoingLinks } from './graph.mjs'
import { RENDER_FORMATS, renderDiagram } from './render.mjs'
import { byCodeUnit, decodeUtf8, excerpt, isPlainObject, parseFailureDetail } from './text.mjs'
import { DestinationError, assertWritableDestination } from './write-guard.mjs'

export { DestinationError, assertWritableDestination } from './write-guard.mjs'

export const TOOL_ID = 'approval-path-visualizer'
export const REPORT_SCHEMA_VERSION = '1'

/**
 * Bounds are part of the contract, not a safety net.
 *
 * An approval graph is ordinary untrusted input: it can be a generated 50 MB
 * file, a thousand steps deep, or a small graph whose edges are arranged to
 * make a path walk quadratic. Every limit below is explicit, overridable from
 * the CLI, and reported by name when it is hit. Exceeding one produces a
 * finding and an `incomplete` report -- never a quietly shorter diagram, and
 * never a pass.
 *
 * `maxTraversalSteps` is a work budget rather than a wall-clock timeout on
 * purpose. A timeout makes the result depend on how busy the machine was, and
 * a report that says "fail" on a fast host and "incomplete" on a slow one is
 * not reproducible. Counting the links the walk follows bounds exactly the
 * same runaway and decides the same way every time.
 */
export const DEFAULT_LIMITS = Object.freeze({
  maxGraphBytes: 1048576,
  maxNodes: 500,
  maxEdgesPerNode: 32,
  maxApprovers: 16,
  maxLabelLength: 120,
  maxDepth: 64,
  maxTraversalSteps: 200000,
})

/**
 * The authoritative rule severity table.
 *
 * Severity is the whole difference between a run that fails and one that
 * passes. Spread across three dozen construction sites as a literal it drifts
 * silently, and flipping one rule down to a warning turns a refusal into a
 * green build with every test still passing. Every finding takes its severity
 * from here, an unknown rule id throws, and `docs/approval-path-rules.md` is
 * asserted against this table in both directions.
 *
 * That is the source of truth, and it is not the guard: agreeing declarations
 * can be edited together, and a test that drives the real binary and then
 * compares against a severity in a map the same edit touches is a fourth
 * declaration, not a guard. So every rule here is pinned again in
 * `test/severity-pins.test.mjs`, which imports no table, reads no catalog and
 * holds no map: one case per rule, every expectation a literal at its
 * assertion -- status, exit code and counted errors for a rule that decides a
 * verdict, and the counted errors plus the severity word the human report
 * prints for one that leaves the run `incomplete`, where the exit code is 2
 * whichever way the rule is graded.
 */
export const RULE_SEVERITY = Object.freeze({
  'approval-cycle': 'error',
  'approvers-duplicate': 'warning',
  'approvers-missing': 'error',
  'approvers-unexpected': 'error',
  'condition-duplicate': 'warning',
  'condition-missing': 'warning',
  'edge-duplicate': 'warning',
  'edge-target-missing': 'error',
  'exception-reason-missing': 'warning',
  'exception-target-missing': 'error',
  'graph-invalid': 'error',
  'graph-key-unknown': 'error',
  'graph-name-missing': 'info',
  'graph-not-json': 'error',
  'graph-not-utf8': 'error',
  'graph-too-large': 'error',
  'graph-unreadable': 'error',
  'label-control-characters': 'warning',
  'label-missing': 'warning',
  'label-too-long': 'error',
  'no-nodes-declared': 'warning',
  'node-id-duplicate': 'error',
  'node-invalid': 'error',
  'node-key-unknown': 'error',
  'node-unreachable': 'error',
  'outcome-missing': 'error',
  'outcome-not-terminal': 'error',
  'outcome-unexpected': 'error',
  'outcome-unreachable': 'error',
  'outcome-value-unknown': 'error',
  'path-dead-end': 'error',
  'path-too-deep': 'error',
  'start-missing': 'error',
  'start-unknown': 'error',
  'timeout-duration-invalid': 'error',
  'timeout-missing': 'warning',
  'timeout-target-missing': 'error',
  'too-many-approvers': 'error',
  'too-many-edges': 'error',
  'too-many-nodes': 'error',
  'traversal-budget-exceeded': 'error',
})

const ALLOWED_OPTIONS = Object.freeze(['format', 'graph', 'limits', 'root'])
const ANCESTOR_PROBE_LIMIT = 64

/**
 * Containment, checked on real paths.
 *
 * Refusing `../` and absolute strings is not confinement: a symbolic link
 * planted inside the declared root resolves out of the tree without ever
 * spelling a traversal. Both sides have been through `realpath` before they
 * reach this function -- comparing a real root against an unresolved candidate
 * refuses legitimate files, which is a bug in the other direction.
 */
export function isInside(root, candidate) {
  if (candidate === root) return true
  return candidate.startsWith(root.endsWith(sep) ? root : `${root}${sep}`)
}

/**
 * Resolve a path that need not exist yet, by realpath-ing its nearest existing
 * ancestor and appending the rest. That still defeats a symlinked parent
 * directory, which is how a "separate destination" quietly becomes a file
 * written back into the input tree.
 */
async function resolveThroughAncestors(absolute) {
  const tail = []
  let probe = absolute
  for (let step = 0; step < ANCESTOR_PROBE_LIMIT; step += 1) {
    try {
      const real = await realpath(probe)
      return tail.length === 0 ? real : join(real, ...tail)
    } catch (error) {
      if (error.code !== 'ENOENT' && error.code !== 'ENOTDIR') {
        throw new TypeError(`Path could not be resolved: ${error.code ?? 'unknown error'}`)
      }
      const parent = dirname(probe)
      if (parent === probe) throw new TypeError('Path has no existing ancestor directory')
      tail.unshift(basename(probe))
      probe = parent
    }
  }
  throw new TypeError('Path is nested too deeply to resolve')
}

/** Resolve the graph file inside the input root. Outside the root is a refusal, not a finding. */
export async function resolveGraphPath(rootReal, relativePath) {
  if (typeof relativePath !== 'string' || relativePath.trim() === '') {
    throw new TypeError('A graph path relative to the root is required')
  }
  if (isAbsolute(relativePath)) {
    throw new TypeError('The graph path must be relative to --root, not absolute')
  }
  const resolved = await resolveThroughAncestors(resolve(rootReal, relativePath))
  if (!isInside(rootReal, resolved)) {
    throw new TypeError('The graph path resolves outside the input root and was refused; nothing was read from it')
  }
  return resolved
}

/**
 * Two names for one file.
 *
 * `realpath` resolves symbolic links, but a hard link has no target to resolve:
 * two names for one inode are two different real paths, so a path comparison
 * says they are different files while a write to either one destroys the other.
 * File identity is the `(device, inode)` pair, and nothing else is. Hard links
 * are ordinary in build trees -- `cp -l`, package stores, backup snapshots --
 * so this is not an exotic case to be waved away.
 */
export async function isSameFile(left, right) {
  let first
  let second
  try {
    first = await stat(left)
  } catch (error) {
    if (error.code === 'ENOENT' || error.code === 'ENOTDIR') return false
    throw new TypeError(`Path could not be inspected: ${error.code ?? 'unknown error'}`)
  }
  try {
    second = await stat(right)
  } catch (error) {
    if (error.code === 'ENOENT' || error.code === 'ENOTDIR') return false
    throw new TypeError(`Path could not be inspected: ${error.code ?? 'unknown error'}`)
  }
  return first.dev === second.dev && first.ino === second.ino
}

/**
 * Resolve the diagram destination and refuse everything that would put it
 * somewhere other than the file the caller named.
 *
 * The diagram is derived from the graph; writing it back into the tree the
 * graph lives in is how a "read-only" tool ends up modifying its own input on
 * the next run. Containment was never the whole of it, and the version of this
 * function that only resolved and compared was measured destroying files:
 *
 * - `--out` pointing at a SYMBOLIC LINK: `resolveThroughAncestors` resolved the
 *   link on its first `realpath` call and handed back the target, so the write
 *   went wherever the link pointed. A 14-byte file outside the root became an
 *   11117-byte SVG while the run exited 0 and printed "diagram written". A link
 *   whose target did not exist yet created the file out there instead.
 * - `--out` under a SYMLINKED PARENT: lexically inside the tree the caller
 *   named, actually outside it, and nothing compared the resolved parent with a
 *   permitted root because there was no permitted root.
 * - `--out` as a HARD LINK to the graph: this one was already refused, and it
 *   stays refused, now by identity inside the shared guard rather than here.
 *
 * `assertWritableDestination` answers all three. Two things are layered on top
 * of it, because the guard cannot know them: the input root, which the diagram
 * must stay out of whatever `--out-root` permits, and the graph file, which is
 * passed as an input so device and inode can see a second name for it.
 */
export async function resolveDiagramDestination(rootReal, destination, graphReal = null, options = {}) {
  if (typeof destination !== 'string' || destination.trim() === '') {
    throw new TypeError('Diagram destination must be a non-empty path')
  }
  const { root = null } = options

  let resolved
  try {
    resolved = await assertWritableDestination(destination, {
      inputs: graphReal === null ? [] : [graphReal],
      root,
      label: '--out',
      rootLabel: '--out-root',
    })
  } catch (error) {
    if (!(error instanceof DestinationError)) throw error
    throw new TypeError(error.message)
  }

  // The destination itself is not a link -- the guard just refused that shape
  // -- so resolving its parent is the whole of the resolution. Containment is
  // then real path against real path in both directions: comparing a resolved
  // destination with an unresolved root refuses every legitimate destination on
  // a host where the input tree sits under a symlinked ancestor.
  const parentReal = await realpath(dirname(resolved))
  if (isInside(rootReal, join(parentReal, basename(resolved)))) {
    throw new TypeError(
      'Diagram destination is inside the input root; the diagram is a derived artifact and must be written elsewhere',
    )
  }
  return resolved
}

export function validateLimits(overrides = {}) {
  if (!isPlainObject(overrides)) throw new TypeError('Limits must be an object')
  const limits = { ...DEFAULT_LIMITS }
  for (const [name, value] of Object.entries(overrides)) {
    if (!Object.hasOwn(DEFAULT_LIMITS, name)) throw new TypeError(`Unknown limit "${name}"`)
    if (!Number.isInteger(value) || value < 1) {
      throw new TypeError(`Limit "${name}" must be a positive integer`)
    }
    limits[name] = value
  }
  return Object.freeze(limits)
}

/**
 * Build one finding, taking its severity from the single table.
 *
 * Exported so a test can prove the refusal below actually throws: a rule that
 * can be emitted without a table entry is a rule whose severity nobody pinned.
 * Every string here has already been through `excerpt`, and goes through it
 * again on the way out -- a message assembled from two sanitised halves is
 * still a message this tool must not be able to forge a line with.
 */
export function createFinding(row) {
  const severity = RULE_SEVERITY[row.ruleId]
  if (severity === undefined) {
    throw new Error(
      `Rule "${row.ruleId}" is not in RULE_SEVERITY; add it to the table and to docs/approval-path-rules.md.`,
    )
  }
  const finding = {
    ruleId: row.ruleId,
    severity,
    message: excerpt(row.message, 400),
    location: { file: excerpt(row.file, 200), pointer: excerpt(row.pointer, 200) },
  }
  if (row.evidence !== undefined && row.evidence !== '') finding.evidence = excerpt(row.evidence, 160)
  if (row.suggestion !== undefined) finding.suggestion = excerpt(row.suggestion, 240)
  return finding
}

function countLinks(nodes, kind) {
  let total = 0
  for (const node of nodes) {
    for (const link of outgoingLinks(node)) if (link.kind === kind) total += 1
  }
  return total
}

function buildReport(rows, incomplete, graph, analysis) {
  rows.sort((left, right) =>
    byCodeUnit(left.file, right.file) ||
    byCodeUnit(left.pointer, right.pointer) ||
    byCodeUnit(left.ruleId, right.ruleId) ||
    byCodeUnit(left.message, right.message))

  const findings = rows.map(createFinding)
  const errors = findings.filter((item) => item.severity === 'error').length
  const warnings = findings.filter((item) => item.severity === 'warning').length
  const nodes = graph === null ? [] : graph.nodes
  const status = incomplete ? 'incomplete' : errors > 0 ? 'fail' : 'pass'

  return {
    schemaVersion: REPORT_SCHEMA_VERSION,
    tool: TOOL_ID,
    status,
    summary: {
      checked: nodes.length,
      errors,
      warnings,
      info: findings.length - errors - warnings,
      approvalSteps: nodes.filter((node) => node.kind === 'approval').length,
      outcomes: nodes.filter((node) => node.kind === 'outcome').length,
      approvers: new Set(nodes.flatMap((node) => node.approvers)).size,
      edges: countLinks(nodes, 'edge'),
      timeouts: countLinks(nodes, 'timeout'),
      exceptionExits: countLinks(nodes, 'exception'),
      unreachable: analysis === null || !analysis.decided
        ? 0
        : nodes.filter((node) => !analysis.layers.has(node.id)).length,
      cycles: analysis === null ? 0 : analysis.cycles.length,
      traversalSteps: analysis === null ? 0 : analysis.steps,
    },
    findings,
  }
}

/**
 * Read one approval graph, check it, and render the diagram.
 *
 * Nothing here reads the network, the clock, the locale or the environment, so
 * two runs over the same bytes produce byte-identical output -- report and
 * diagram alike. The returned `diagram` is a string the caller decides what to
 * do with; this function writes nothing.
 */
export async function visualizeApprovalPath(options = {}) {
  if (!isPlainObject(options)) throw new TypeError('Options must be an object')
  for (const key of Object.keys(options)) {
    if (!ALLOWED_OPTIONS.includes(key)) throw new TypeError(`Unknown option "${key}"`)
  }
  if (typeof options.root !== 'string' || options.root.trim() === '') {
    throw new TypeError('An input root is required')
  }
  const format = options.format ?? 'svg'
  if (!RENDER_FORMATS.includes(format)) {
    throw new TypeError(`Unknown format "${excerpt(String(format), 40)}"; expected one of ${RENDER_FORMATS.join(', ')}`)
  }
  const limits = validateLimits(options.limits ?? {})

  let rootReal
  try {
    rootReal = await realpath(resolve(options.root))
  } catch (error) {
    throw new TypeError(`Input root could not be read: ${error.code ?? 'unknown error'}`)
  }
  const rootInfo = await stat(rootReal)
  if (!rootInfo.isDirectory()) throw new TypeError('Input root must be a directory')

  const graphReal = await resolveGraphPath(rootReal, options.graph)
  const file = excerpt(String(options.graph).split(sep).join('/'), 200)

  const rows = []
  let incomplete = false
  const record = (row) => {
    rows.push({ file, ...row })
    if (row.incomplete === true) incomplete = true
  }

  /** Everything below reads; the only thing that ever leaves is a string. */
  const unreadable = (message, suggestion) => {
    record({ ruleId: 'graph-unreadable', pointer: '/', message, suggestion, incomplete: true })
    return { report: buildReport(rows, incomplete, null, null), diagram: null, graph: null, analysis: null, file }
  }

  let info
  try {
    info = await stat(graphReal)
  } catch (error) {
    return unreadable(
      `Graph file could not be read: ${error.code ?? 'unknown error'}.`,
      'Check the path given to --graph, relative to --root.',
    )
  }
  if (!info.isFile()) {
    return unreadable('Graph path is not a regular file, so nothing could be read from it.')
  }
  if (info.size > limits.maxGraphBytes) {
    record({
      ruleId: 'graph-too-large',
      pointer: '/',
      message: `Graph file is ${info.size} bytes, above the maxGraphBytes limit of ${limits.maxGraphBytes}; it was not parsed.`,
      suggestion: 'Raise --max-graph-bytes, or split the approval path into separate graphs.',
      incomplete: true,
    })
    return { report: buildReport(rows, incomplete, null, null), diagram: null, graph: null, analysis: null, file }
  }

  let bytes
  try {
    bytes = await readFile(graphReal)
  } catch (error) {
    return unreadable(`Graph file could not be read: ${error.code ?? 'unknown error'}.`)
  }

  const decoded = decodeUtf8(bytes)
  if (!decoded.ok) {
    record({
      ruleId: 'graph-not-utf8',
      pointer: '/',
      message: 'Graph file is not valid UTF-8; it was not parsed and nothing was drawn from it.',
      suggestion: 'Re-encode the graph as UTF-8.',
      incomplete: true,
    })
    return { report: buildReport(rows, incomplete, null, null), diagram: null, graph: null, analysis: null, file }
  }

  let value
  try {
    value = JSON.parse(decoded.text)
  } catch (error) {
    record({
      ruleId: 'graph-not-json',
      pointer: '/',
      message: 'Graph file is not valid JSON; it was not parsed and nothing was drawn from it.',
      evidence: excerpt(parseFailureDetail(error), 120),
      incomplete: true,
    })
    return { report: buildReport(rows, incomplete, null, null), diagram: null, graph: null, analysis: null, file }
  }

  const compiled = compileGraph(value, limits)
  for (const problem of compiled.problems) record(problem)

  if (compiled.graph === null) {
    return { report: buildReport(rows, incomplete, null, null), diagram: null, graph: null, analysis: null, file }
  }

  /**
   * Green on no evidence is a defect, not a clean bill of health. A graph that
   * declared no usable step means the run obtained nothing to judge, so it is
   * reported and the run is incomplete -- it can never be a pass.
   */
  if (compiled.graph.nodes.length === 0) {
    record({
      ruleId: 'no-nodes-declared',
      pointer: '/nodes',
      message: 'No approval step was read from this graph, so this run checked nothing and drew nothing.',
      suggestion: 'Declare the approval steps in "nodes".',
      incomplete: true,
    })
    return { report: buildReport(rows, incomplete, null, null), diagram: null, graph: null, analysis: null, file }
  }

  const analysis = analyseGraph(compiled.graph, limits)
  for (const problem of analysis.problems) record(problem)

  const report = buildReport(rows, incomplete, compiled.graph, analysis)
  const diagram = renderDiagram({
    graph: compiled.graph,
    analysis,
    report,
    format,
    title: basename(file),
  })

  return { report, diagram, graph: compiled.graph, analysis, file }
}

/** Write the derived diagram to its own destination. The input is never touched. */
export async function writeDiagram(destination, diagram) {
  await mkdir(dirname(destination), { recursive: true })
  await writeFile(destination, diagram, 'utf8')
  return destination
}

const SEVERITY_WIDTH = 7

export function formatReport(report, { file, name = '', diagram = null } = {}) {
  const { summary } = report
  const subject = name === '' ? file : `${name} (${file})`
  const lines = [
    `${subject}: ${summary.approvalSteps} approval step(s), ${summary.outcomes} outcome(s), ` +
    `${summary.approvers} approver(s), status ${report.status}.`,
    `paths: ${summary.edges} edge(s), ${summary.timeouts} timeout(s), ` +
    `${summary.exceptionExits} exception exit(s), ${summary.unreachable} unreachable step(s), ` +
    `${summary.cycles} cycle(s).`,
  ]
  if (diagram !== null) lines.push(`diagram: ${excerpt(diagram, 200)}`)
  for (const finding of report.findings) {
    lines.push(
      `${finding.severity.toUpperCase().padEnd(SEVERITY_WIDTH)} ` +
      `${finding.location.file}${finding.location.pointer} ${finding.ruleId} ${finding.message}`,
    )
  }
  return `${lines.join('\n')}\n`
}

export {
  analyseGraph,
  canonicalCycle,
  compileGraph,
  findCycles,
  layerFromStart,
  outgoingLinks,
  EDGE_KEYS,
  EXCEPTION_KEYS,
  GRAPH_KEYS,
  NODE_KEYS,
  NODE_KINDS,
  OUTCOME_VALUES,
  SUPPORTED_SCHEMA_VERSIONS,
  TIMEOUT_KEYS,
} from './graph.mjs'
export { RENDER_FORMATS, headerLines, layoutGraph, renderDiagram } from './render.mjs'
export {
  byCodeUnit,
  decodeUtf8,
  escapeMarkup,
  excerpt,
  hasControlCharacters,
  isIdentifier,
  markupText,
  parseDuration,
  parseFailureDetail,
  wrapText,
} from './text.mjs'
